File: 31 ObjC/31 ObjC/ViewController.m
//
//  ViewController.m
//  31 ObjC
//
//  Created by Johan Wigmo on 2019-01-31.
//  Copyright © 2019 Newton. All rights reserved.
//

#import "ViewController.h"
#import "Person.h"
#import "_1_ObjC-Swift.h"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ViewController {
    
    NSMutableArray<Person *> *people;
    
}

@synthesize test;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    
    [self populate];
    
    Work *work = [[Work alloc] initWithName:@"Name"];
    NSString *name = work.name;
    NSLog(@"Work name: %@", name);
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return people.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"cell"];
    }
    
    Person *person = people[indexPath.row];
    
    cell.textLabel.text = person.name;
    
    NSString *detail = [NSString stringWithFormat:@"%@ (%i)", person.nickname, person.age];
    cell.detailTextLabel.text = detail;
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

-(void)populate
{
    people = [[NSMutableArray alloc] init];
    
    for (int i = 0; i<30; i++) {
        Person *person = [[Person alloc] initWithName:@"Kalle" age:i*6 andNickname:@"Pettersson"];
        [people addObject:person];
    }
    
    [self.tableView reloadData];
}



@end

File: 28 Test/28 Test/28 Test/ViewController.swift
//
//  ViewController.swift
//  28 Test
//
//  Created by Johan Wigmo on 2019-01-28.
//  Copyright © 2019 Newton. All rights reserved.
//

import UIKit

class ViewController: UIViewController, DataProviderDelegate {

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var textField: UITextField!
    @IBOutlet weak var addButton: UIButton!
    
    var dataProvider: DataProvider?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if dataProvider == nil {
            dataProvider = DataProvider(manager: DataManager())
        }
        
        dataProvider?.delegate = self
        
        tableView.dataSource = dataProvider
        tableView.delegate = dataProvider
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        
    }

    @IBAction func add() {
        guard let text = textField.text else { return }
        textField.text = ""
        dataProvider?.manager.add(text: text)
        textField.resignFirstResponder()
        tableView.reloadData()
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let destination = segue.destination as? SecondViewController, let text = sender as? String {
            destination.text = text
        }
    }
    
    func didSelect(text: String) {
        performSegue(withIdentifier: "showSecond", sender: text)
    }
    
}


